server/src/main/java/io/mantisrx/master/jobcluster/job/JobActor.java

```diff
@@ -1380,13 +1395,18 @@ private void initializeRunningWorkers() {
                                     wm.getWorkerIndex(), wm.getWorkerNumber(), Status.TYPE.HEARTBEAT, "",
                                     MantisJobState.Started, System.currentTimeMillis()));
                             worker.processEvent(fakeHB, jobStore);
-                        } catch (Exception e) {
+                        } catch (MissingWorkerPortsException e) {
+                            numMissingWorkerPorts.increment();
+                            // Mark these workers with missing ports for resubmission.
+                            workersToResubmit.add(worker);
+                        } catch (InvalidWorkerStateChangeException | IOException e) {
                             e.printStackTrace();
                         }
                         workerHosts.put(wm.getWorkerNumber(),
-                                new WorkerHost(wm.getSlave(), wm.getWorkerIndex(),
-                                        wm.getPorts().isPresent() ? wm.getPorts().get().getPorts()
-                                                : Lists.newArrayList(),
+                                new WorkerHost(
+                                        wm.getSlave(),
+                                        wm.getWorkerIndex(),
+                                        wm.getPortList(),
```

Comment:
Need a review here too. The previous boolean condition is now delegating logic to the `MantisWOrkerMetadataImpl`. It does the same thing, except it returns a list of ports (good or bad) instead of an empty list. Intent is to not get array index out of bounds exception.

Reply:
lgtm